checker-util/src/main/java/org/checkerframework/checker/regex/util/RegexUtil.java

```diff
@@ -334,4 +427,159 @@ private static String regexErrorMessage(String s, int expectedGroups, int actual
+        ArrayDeque<Integer> unclosedCapturingGroups = new ArrayDeque<>();
+
+        // Stores whether the last occurrence of '(' in the regex marked a capturing group (true) or
+        // a non-capturing group (false).
+        // Element is pushed to it when a '(' is encountered. If it is a capturing group (named or
+        // unnamed), push true otherwise push false.
+        // Element is popped from it when a ')' is encountered. If the popped element is true, it
+        // means that the ')' closes a capturing group (we need to pop from the Integer deque)
+        // otherwise it closes a non-capturing group.
+        ArrayDeque<Boolean> lastUnclosedWasCapturingGroup = new ArrayDeque<>();
+
+        // Number of capturing groups encountered. Used for numbering of the capturing groups.
+        // Increments every time a '(' that opens a capturing group is encountered during the
+        // traversal. This is the value that is pushed onto the Integer deque.
+        int group = 0;
+
+        // Optional group here onwards means the ith capturing group, which may not match any part
+        // of a text that matched the regular expression and thus may return null on calls to
+        // matcher.group(i).
+
+        // If you encounter '(', check the next character. If it is a '?', it is a special
+        // construct,
+        // (either pure, non-capturing groups that do not capture text and do not count towards the
+        // group total, or named-capturing group) and we need to check the next character. If the
+        // character following the '?' is a '<' the '(' represents the opening of a named-capturing
+        // group and it will be handled like a normal capturing group. If the '(' was not followed
+        // by a '?', it is a normal capturing group.
+        // In case of capturing groups, increment the group variable and push it to the
+        // unclosedCapturingGroups deque. Push true to the lastUnclosedWasCapturingGroup deque.
+        // In case of non-capturing groups, push false to the lastUnclosedWasCapturingGroup deque.
+        // We need the boolean deque so that when we encounter a ')', we can know whether it closes
+        // a capturing group (the top element is true) or a non-capturing group (the top element is
+        // false). One additional check is required. If '(' represented a capturing a group and was
+        // preceded by '|', it is an optional group.
+
+        // If you encounter ')', check the top of the lastUnclosedWasCapturingGroup deque. If it was
+        // false, do nothing otherwise remove the top element from the deque since it is now closed
+        // and check if it is followed by a '?', '*', '|' or '{0'. If it is, then it is an optional
+        // group otherwise not. If it is an optional group, remove it from the nonNullGroups list.
+
+        // If you encounter '[', traverse the regex till you find the closing '[', you may encounter
```

Comment:
"closing ']'" should be "closing '['".  However, I think the code will be easier to read with comments like this inline rather than a huge comment at the beginning, which prevents readers from seeing which parts of the comment correspond to which parts of the code.

Reply:
I'll inline the big wall of comment. It is taking some time. I'll push my changes by tomorrow.